### ISSUE TITLE
NestJS class-validator is not the same in e2e testing mode

### ISSUE THREAD

**Issue by kykungz** [nestjs/nest#5264]
## Bug Report

Seems like NestJS Class Validation is not behaving the same in testing environment (with `supertest`) and in dev/prod environment (`axios`, Postman).

<img width="548" alt="Screen Shot 2563-08-13 at 02 49 56" src="https://user-images.githubusercontent.com/9304909/90060967-af172900-dd0f-11ea-8b73-678b46a80b6c.png">

With the above setting, every `BaseDto`'s property is required, and every `PartialDto` is optional.

This works correctly when requesting to NestJS application in dev/prod mode through tools such as `axios`, `curl` or Postman.

However, when testing it through NestJS's e2e testing guide (eg. https://docs.nestjs.com/fundamentals/testing#end-to-end-testing), it appears to not behave the same as above.

My guess is that `PartialType` is not working correctly in testing mode. Thus, making `PartialDto` not partially required as it should be (TypeScript compilation problem?). 

## Current behavior

When requesting to endpoint with `PartialDto`, sending only `firstName` will be fine and returned 200.

However, doing the same with e2e testing will fail with the following error response:
```
 '{"statusCode":400,"message":["firstName must be a string"],"error":"Bad Request"}'
```

## Input Code

Here is the reproduction Repo: https://github.com/kykungz/nest-test-issue

Try sending a request by running
```sh
yarn test:e2e
```

and compare it with using `curl`
```sh
curl --location --request POST 'http://localhost:5000' \
--header 'Content-Type: application/json' \
--data-raw '{
    "firstName": "test2@test.com"
}'
```

after starting the app (`yarn start:dev`)

## Expected behavior

The e2e test should pass just like Postman and curl.

`PartialDto` should accept partial properties as it should be.

## Environment

<pre><code>
Nest version: 7.4.1

 
For Tooling issues:
- Node version: v10.20.1 
- Platform:  Mac

Others:

</code></pre>

**Comment by squid17** [comment]
Try just pass it manually in your e2e file

```
app = moduleFixture.createNestApplication();
app.useGlobalPipes(new ValidationPipe());
await app.init();
```

This works for me.

**Comment by kamilmysliwiec** [comment]
Just tested and it works fine

**Comment by lusatiro** [comment]
How could make the Query works in e2e?

**Comment by guitoof** [comment]
Hi there, I encountered the exact same issue while following the [NestJS Course](https://learn.nestjs.com/courses).
Great job on the course and thanks by the way, I'm a big fan of NestJS ;)


TLDR: it works fine when you `import { PartialType } from '@nestjs/mapped-typed'` but fails when you `import { PartialType from '@nestjs/swagger' => If you're using PartialType from @nestjs/swagger, make sure all your properties are decorated with the @ApiProperty decorator`
@kykungz, I hope this can help you

In the ["Validate Input Data with Data Transfer Object"](https://learn.nestjs.com/courses/591712/lectures/23192389) section of the course, we use PartialType from `@nestjs/mapped-type`.
Later on, in the ["Decorating Model Properties"](https://learn.nestjs.com/courses/591712/lectures/23275310)  section, we switch to use PartialType from `@nestjs/swagger`
While the actual request made with CURL (or Postman ...) still works fine with a partial input:
```
{
	"name": "Coffee Name"
}
```
The e2e Test fails with a 400 error :
```
{
      statusCode: 400,
      message: [ 'each value in flavors must be a string' ],
      error: 'Bad Request',
      timestamp: '2021-01-30T21:19:39.553Z'
}
```
In my case, this error only occurs on the `flavors` property (see create-coffee.dto.ts file below) on which I haven't add the @ApiProperty decorator. 
As you can see in the e2e test file above, I use the ValidationPipe in my Test app as well.

[coffee.dto.zip](https://github.com/nestjs/nest/files/5898821/coffee.dto.zip)
[coffees.e2e-spec.ts.zip](https://github.com/nestjs/nest/files/5898810/coffees.e2e-spec.ts.zip)

@kamilmysliwiec it seems like the PartialType from @nestjs/swagger only succeed to make a property "Partial" when it has the @ApiProperty decorator. It feels coincidential,  is this on purpose ?  If so would you mind explaining it ?
Thank you very much for your response.

**Comment by abramyk-mv** [comment]
@guitoof Thanks a bunch for your research!

Using `PartialType` from `@nestjs/swagger` along with `@ApiProperty` decorator really made my e2e tests work.

But it breaks the original DTO because `@nestjs/swagger/plugin` (which I assume you use) [automatically adds ApiProperty decorator](https://docs.nestjs.com/openapi/types-and-parameters).
After some researches, I discovered that [Nest does not use CLI compiler and does not apply any plugins](https://docs.nestjs.com/openapi/cli-plugin#integration-with-ts-jest-e2e-tests) (including `@nestjs/swagger/plugin`) to run e2e tests.
So, we should try to enable the plugin manually to make it work as expected.

**Comment by jmcdo29** [comment]
There's a [section in the docs](https://docs.nestjs.com/graphql/cli-plugin#integration-with-ts-jest-e2e-tests) about integrating the `@nestjs/graphql/plugin` with `ts-jest`. You could probably follow the same steps here and get the integration working properly.

**Comment by kamilmysliwiec** [comment]
@jmcdo29 there's a section about integrating the `@nestjs/swagger/plugin` too https://docs.nestjs.com/openapi/cli-plugin#integration-with-ts-jest-e2e-tests :)

**Comment by abramyk-mv** [comment]
@kamilmysliwiec It didn't work for me, `@nestjs/swagger` `PartialType` was still ignored in my e2e tests.

But still, I've managed to make `@nestjs/swagger/plugin` work properly using the steps provided in the [article](https://dev.to/nestjs/using-the-nestjs-cli-graphql-plugin-for-e2e-testing-4dpb) posted by @jmcdo29

**Comment by guitoof** [comment]
> @jmcdo29 there's a section about integrating the `@nestjs/swagger/plugin` too https://docs.nestjs.com/openapi/cli-plugin#integration-with-ts-jest-e2e-tests :)

@kamilmysliwiec thanks for your response. This helped me integrate the swagger plugin to be used by ts-jest (and understood why it did not work as expected in the first place.
I think it would be a good idea to add this information to the NestJS Course to make sure we go through [this documentation](https://docs.nestjs.com/openapi/cli-plugin#integration-with-ts-jest-e2e-tests) in the e2e section if we use the swagger API.
Anyway, thanks again for the help !

FYI: I had to upgrade `ts-jest` from 26.1.0 to 26.3.0 for the code snippets for the `jest-e2e.json`  to work
